src/mlpack/methods/kde/kde_rules_impl.hpp

```diff
@@ -0,0 +1,130 @@
+ *
+ * Implementation of rules for Kernel Density Estimation with generic trees.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+
+#ifndef MLPACK_METHODS_KDE_RULES_IMPL_HPP
+#define MLPACK_METHODS_KDE_RULES_IMPL_HPP
+
+// In case it hasn't been included yet.
+#include "kde_rules.hpp"
+
+namespace mlpack {
+namespace kde {
+
+template<typename MetricType, typename KernelType, typename TreeType>
+KDERules<MetricType, KernelType, TreeType>::KDERules(
+    const arma::mat& referenceSet,
+    const arma::mat& querySet,
+    arma::vec& densities,
+    const double error,
+    const std::vector<size_t>& oldFromNewQueries,
+    MetricType& metric,
+    const KernelType& kernel) :
+    referenceSet(referenceSet),
+    querySet(querySet),
+    densities(densities),
+    error(error),
+    oldFromNewQueries(oldFromNewQueries),
+    metric(metric),
+    kernel(kernel),
+    lastQueryIndex(querySet.n_cols),
+    lastReferenceIndex(referenceSet.n_cols),
+    baseCases(0),
+    scores(0)
+{
+  // Nothing to do.
+}
+
+//! The base case.
+template<typename MetricType, typename KernelType, typename TreeType>
+inline force_inline
+double KDERules<MetricType, KernelType, TreeType>::BaseCase(
+    const size_t queryIndex,
+    const size_t referenceIndex)
+{
+  double distance = metric.Evaluate(querySet.col(queryIndex),
+                                    referenceSet.col(referenceIndex));
+  densities(oldFromNewQueries.at(queryIndex)) += kernel.Evaluate(distance);
+  ++baseCases;
+  lastQueryIndex = queryIndex;
+  lastReferenceIndex = referenceIndex;
+  return distance;
+}
+
+//! Single-tree scoring function.
+template<typename MetricType, typename KernelType, typename TreeType>
+double KDERules<MetricType, KernelType, TreeType>::
+Score(const size_t /* queryIndex */, TreeType& /* referenceNode */)
+{
+  ++scores;
+  traversalInfo.LastScore() = 0.0;
+  return 0.0;
+}
+
+template<typename MetricType, typename KernelType, typename TreeType>
+double KDERules<MetricType, KernelType, TreeType>::Rescore(
+    const size_t /* queryIndex */,
+    TreeType& /* referenceNode */,
+    const double oldScore) const
+{
+  // If it's pruned it continues to be pruned.
+  return oldScore;
+}
+
+//! Double-tree scoring function.
+template<typename MetricType, typename KernelType, typename TreeType>
+double KDERules<MetricType, KernelType, TreeType>::
+Score(TreeType& queryNode, TreeType& referenceNode)
+{
+  double score, bound;
+  bound = kernel.Evaluate(queryNode.MinDistance(referenceNode)) -
+          kernel.Evaluate(queryNode.MaxDistance(referenceNode));
+
+  if (bound <= error / referenceSet.n_cols)
+  {
+    arma::vec queryCenter, referenceCenter;
+    referenceNode.Center(referenceCenter);
+    queryNode.Center(queryCenter);
```

Comment:
Here we could actually use the tree traits for optimization.  For some tree types, the first point held in the tree is also the center of the node, so we could do

```
if (TreeTraits<TreeType>::FirstPointIsCentroid)
{
  // use queryNode.Dataset().col(queryNode.Point(0)) instead of queryCenter, and
  // referenceNode.Dataset().col(referenceNode.Point(0)) instead of referenceCenter.
}
else
{
  // use queryCenter and referenceCenter
}
```

This could help save a lot of time for some tree types.

Reply:
I understand, but doesn't it make more sense that the tree handles this?
Currently cover tree (and its derivates) is the only one that has ```FirstPointIsCentroid``` and it does a good job at handling right?

```C++
 //! Get the center of the node and store it in the given vector.
  void Center(arma::vec& center) const
  {
    center = arma::vec(dataset->col(point));
  }
...
...
...
  //! Index of the point in the matrix which this node represents.
  size_t point;
```

Although it's easy to implement, maybe for future cases, but I just had this doubt :smile:

Reply:
Yes, but the problem is that when you call `CoverTree::Center()` it will copy the column of the dataset.  If you use `FirstPointIsCentroid`, we can avoid the copy, and this can be a noticeable acceleration.